flask.ext Import Migration Tool #1135

keyanp opened the pull request:
This is an early commit to get some feedback. This script supports single imports in two formats. I am concerned that the naive approach is not what you guys are looking for. I would obvious add support for multiple imports and the third case `import flask.ext.foo` => `import flask_foo`. But the solution is definitely super hacky.

untitaker commented:
>This script supports single imports in two formats. I am concerned that the naive approach is not what you guys are looking for.

Yeah... There are some forms with which this script would break:

    from flask.ext.foo import \
        a, \
        b

    from flask.ext.foo import (
        a,
        b
    )

Parsing definetly should be delegated to some third-party library. As said in #1135, https://github.com/python-modernize/python-modernize looks great, but I am not sure if one can easily extend it with own fixers.

keyanp commented:
I looked into using lib2to3 initially but I didn't look at python-modernize. Upon inspection I think writing a custom fixer wouldn't be too bad. I'll work on this over the weekend. Thanks for the feedback.

untitaker commented:
@keyanp Just found https://github.com/Psycojoker/baron, maybe you'll find it useful.

keyanp commented:
@untitaker Thanks! That looks promising, I've been fiddling with writing a fixer for lib2to3, but it's pretty confusing. Hopefully this will be easier.

keyanp commented:
@untitaker So RedBaron ended up being perfect for the job, the new module supports everything mentioned in issue #1135 except it doesn't change imports in the format:

`import flask.ext.foo.[more stuff]`

it only handles:

`import flask.ext.foo`

And even it that case, it doesn't fix every function call in the source. I also added a few tests, but RedBaron needs to be installed on Travis in order to pass. Let me know how you want me to handle the above issue.

untitaker commented on scripts/flaskext_migrate.py:
```diff
@@ -0,0 +1,94 @@
+    Case 2: from flask.ext import foo --> import flask_foo as foo
+    """
+    from_imports = red.find_all("FromImport")
+    for x in range(len(from_imports)):
```
You can use ``enumerate`` here.

untitaker commented:
Regarding renaming functions, I think finding all ``NameNode`` instances and replacing them will succeed for most cases (although it might refactor code that doesn't need it, but I don't think we should care about that). That would probably also make it easier to handle the other import type properly.

There's an edge case that your tests don't cover yet, although I don't know if your code does:

    from flask.ext.foo import bar as baz

That said, this looks *really* nice! Apart from the inline code comments I gave, please also add a simple usage message if the user hasn't provided ``sys.argv[1]``.

keyanp commented:
@untitaker So I added a test for that case and fixed the logic to cover it. Also added a message if no argument is provided. I still need to fix another edge case, I should push that soon. Just a couple things I have questions about:
1. How do I pip install RedBaron on the Travis build so I can pass the build?
2. What is the correct way to import the script into my tester file? I'm just using: `sys.path.append('scripts')` right now, which works locally, but I don't think it will work when building on Travis.

untitaker commented:
>1. How do I pip install RedBaron on the Travis build so I can pass the build?

Enter it as dependency directly after ``greenlet`` under ``deps= ...``

>What is the correct way to import the script into my tester file?

I was about to suggest that you move the tests for this script to the same directory as the script itself. py.test should pick it up automatically.

keyanp commented:
Great! Thanks!

keyanp commented:
So I moved the test file, added RedBaron to the dependencies, and also added an extra test for this case (which is now covered):

>from flask.ext.foo import (
>    a,
>  b
>)

untitaker commented:
Yeah, we still have to fix imports like these:

    import flask.ext.foo

Just warning about it is unfortunately not an option. :(

keyanp commented:
Yeah no problem, could you refer to the specific case you are talking about? The case you wrote in the last comment already works: import flask.ext.foo --> import flask_foo

untitaker commented:
I meant rewriting the usages of ``flask.ext.foo`` to the new form inside the code itself. See my comment above about NameNode.

keyanp commented:
@untitaker So I couldn't figure out a good way to do it using the NameNodes, but the function calls are represented by Atomtrailers and replacing those worked for some test files I ran it against. I added a simple test to cover that case as well. Otherwise I refactored the code a bit.

untitaker commented on scripts/flaskext_migrate.py:
```diff
@@ -0,0 +1,147 @@
+        try:
+            if (node.value[0].value == 'flask' and
+               node.value[1].value == 'ext'):
+                node.replace("flask_foo%s" % node.value[3])
```
I think that's a leftover, "foo" is not supposed to be hardcoded. :)

keyanp replied:
Woops, good catch. Fixed that.

untitaker commented on scripts/test_import_migration.py:
```diff
@@ -0,0 +1,63 @@
+def test_function_call_migration():
+    red = RedBaron("flask.ext.foo(var)")
+    output = migrate.fix_tester(red)
+    assert output == "flask_foo(var)"
```
This works, but this doesn't:

    def test_function_call_migration():
        red = RedBaron("import flask.ext.foo\n\n"
                       "flask.ext.foo.bar(var)")
        output = migrate.fix_tester(red)
        assert output == ("import flask_foo\n\n"
                          "flask_foo.bar(var)")

keyanp replied:
Okay, there are a few problems with this. Let me work on it. Thanks for looking it over.

keyanp commented:
@untitaker Refactored to clean it up a bit, added the test you mentioned, and added a function to cover it. Thanks again for your help.

DasIch commented:
Looks good to me.

DasIch commented:
@untitaker What's your opinion?

untitaker commented:
In IRC I expressed the concern that this tool might rewrite things that it is not supposed to rewrite, because it doesn't do a lot of "checks" on the AST.

But since this is just a feeling I have, this is actually pretty great. Thanks @keyanp!

davidism commented:
This moved to its own repo: [pocoo/flask-ext-migrate](https://github.com/pocoo/flask-ext-migrate).  This should probably be removed from the Flask repo and the documentation should be updated to mention the migration tool.

untitaker commented:
Agreed, please file a new issue with a blocker label and appropriate milestone.

On 4 April 2016 20:23:10 CEST, David Lord <notifications@github.com> wrote:
>This moved to its own repo:
>[pocoo/flask-ext-migrate](https://github.com/pocoo/flask-ext-migrate). 
>This should probably be removed from the Flask repo and the
>documentation should be updated to mention the migration tool.
>
>---
>You are receiving this because you modified the open/close state.
>Reply to this email directly or view it on GitHub:
>https://github.com/pallets/flask/pull/1342#issuecomment-205433643

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

untitaker commented:
For future reference, followup on https://github.com/pallets/flask/issues/1765
